src/Bicep.LangServer/Refactor/ExpressionAndTypeExtractor.cs

```diff
@@ -345,45 +368,131 @@ private string GetQuotedText(string text)
+
+        for (int line = statementStartLine; line >= 1; --line)
+        {
+            var (hasContent, hasComments) = CheckLineContent(lineStarts, programSyntax, line - 1);
+            if (hasComments && !hasContent)
+            {
+                continue;
+            }
+            else
+            {
+                return line;
             }
         }
 
-        // If no existing declarations of the desired type, insert right before the statement containing the extraction expression
-        return GetOffset(extractionLine, 0);
+        return 0;
+    }
+
+    private static int GetLastLineOfStatement(CompilationContext compilationContext, StatementSyntax statementSyntax)
+    {
+        return TextCoordinateConverter.GetPosition(compilationContext.LineStarts, statementSyntax.GetEndPosition()).line;
+    }
+
+    private (bool addBefore, bool addAfter) ShouldAddBlankLines(StatementSyntax statementSyntax)
+    {
+        bool addBefore = false;
+        bool addAfter = false;
+
+        var startingLine = GetFirstLineOfStatementIncludingComments(compilationContext, statementSyntax);
+        var endingLine = GetLastLineOfStatement(compilationContext, statementSyntax);
+
+        bool? hasBlankLineBefore = IsFirstLine(startingLine) ? null : CheckLineContent(startingLine - 1).IsEmpty;
+        bool? hasBlankLineAfter = IsLastLine(endingLine) ? null : CheckLineContent(endingLine + 1).IsEmpty;
+
+        bool existingDeclarationUsesBlankLines = hasBlankLineBefore ?? hasBlankLineAfter ?? true;
+        addBefore = existingDeclarationUsesBlankLines;
+        addAfter = hasBlankLineAfter ?? true;
+
+        // Don't add another after if there's already one
+        addAfter = addAfter && (IsLastLine(endingLine) || CheckLineContent(endingLine + 1).IsNotEmpty);
+
+        return (addBefore, addAfter);
+    }
 
-        StatementSyntax? StatementOfTypeAtLine(int line)
+    private sealed class CheckContentVisitor : CstVisitor
+    {
+        private readonly TextSpan span;
+
+        public CheckContentVisitor(TextSpan span)
         {
-            var lineOffset = GetOffset(line, 0);
-            var statementAtLine = SyntaxMatcher.FindNodesSpanningRange(compilationContext.ProgramSyntax, lineOffset, lineOffset)
-                .OfType<StatementSyntax>()
-                .FirstOrDefault(s => s.GetType() == declarationSyntaxType);
-            return statementAtLine;
+            this.span = span;
         }
 
-        (int line, int character) GetPosition(int offset) => TextCoordinateConverter.GetPosition(lineStarts, offset);
-        int GetOffset(int line, int character) => TextCoordinateConverter.GetOffset(lineStarts, line, character);
+        public bool HasContent { get; private set; } = false;
+        public bool HasComments { get; private set; } = false;
+        public bool HasContentOrComments => HasContent || HasComments;
+
+        public override void VisitSyntaxTrivia(SyntaxTrivia syntaxTrivia)
+        {
```

Comment:
We can bail out if `HasContent` and `HasComments` are already true to avoid visiting other syntax nodes.

Reply:
Did that, plus an extra optimization, please review: 643f3404afddef0b59eae6b573fb2ef65e903572